src/storage/v2/indices/text_index.cpp

```diff
@@ -0,0 +1,233 @@
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "storage/v2/indices/text_index.hpp"
+#include "query/db_accessor.hpp"
+#include "storage/v2/view.hpp"
+#include "text_search.hpp"
+
+namespace memgraph::storage {
+
+void TextIndex::AddNode(Vertex *vertex_after_update, Storage *storage,
+                        const std::vector<mgcxx::text_search::Context *> &applicable_text_indices) {
+  // NOTE: Text indexes are presently all-property indices. If we allow text indexes restricted to specific properties,
+  // an indexable document should be created for each applicable index.
+  nlohmann::json document = {};
+  nlohmann::json properties = {};
+  for (const auto &[prop_id, prop_value] : vertex_after_update->properties.Properties()) {
+    if (!prop_value.IsString()) continue;
+    properties[storage->PropertyToName(prop_id)] = prop_value.ValueString();
+  }
+
+  document["data"] = properties;
+  document["metadata"] = {};
+  document["metadata"]["gid"] = vertex_after_update->gid.AsInt();
+  // TODO add txid
+  document["metadata"]["deleted"] = false;
+  document["metadata"]["is_node"] = true;
+
+  for (auto *index_context : applicable_text_indices) {
+    try {
+      mgcxx::text_search::add_document(*index_context, mgcxx::text_search::DocumentInput{.data = document.dump()},
+                                       false);
+    } catch (const std::exception &e) {
+      throw query::QueryException(fmt::format("Tantivy error: {}", e.what()));
+    }
+  }
+}
+
+void TextIndex::AddNode(Vertex *vertex_after_update, Storage *storage) {
+  auto applicable_text_indices = GetApplicableTextIndices(vertex_after_update);
+  if (applicable_text_indices.empty()) return;
+  AddNode(vertex_after_update, storage, applicable_text_indices);
+}
+
+void TextIndex::UpdateNode(Vertex *vertex_after_update, Storage *storage) {
+  auto applicable_text_indices = GetApplicableTextIndices(vertex_after_update);
+  if (applicable_text_indices.empty()) return;
+  RemoveNode(vertex_after_update, applicable_text_indices);
+  AddNode(vertex_after_update, storage, applicable_text_indices);
+}
+
+void TextIndex::UpdateNode(Vertex *vertex_after_update, Storage *storage, const std::vector<LabelId> &removed_labels) {
+  auto indexes_to_remove_node_from = GetApplicableTextIndices(removed_labels);
+  RemoveNode(vertex_after_update, indexes_to_remove_node_from);
+
+  auto indexes_to_update_node = GetApplicableTextIndices(vertex_after_update);
+  if (indexes_to_update_node.empty()) return;
+  RemoveNode(vertex_after_update, indexes_to_update_node);
+  AddNode(vertex_after_update, storage, indexes_to_update_node);
+}
```

Comment:
This is tricky because if we just delete a node, in snapshot isolation, accessing index will return wrong data -> figure out how to fix in Phase#1 (not strickly under this PR)

Reply:
Snapshot isolation is now deferred to Phase 4 of the project (not in this PR/release)